File: LogiOS/TrustarApp/Services/Request/UnscheduledBin.swift
//
//  UnscheduledBin.swift
//  TrustarApp
//
//  Created by CuongNguyen on 10/03/2022.
//  Copyright © 2022 DionSoftware. All rights reserved.
//

import Foundation

class UnscheduledBin: Codable {
    var companyCd: String
    var userId: String
    var truckCd: String
    var latitude: Double
    var longitude: Double
    var accuracy: Double
    var executeDate: Int64
    var outgoingMeter: Int? = nil
    var allocationToken: AllocationToken
    var clientInfo: ClientInfo
    
    init(token: String,
         truck: Truck,
         locationRecord: LocationRecord,
         clientInfo: ClientInfo,
         userInfo: UserInfo,
         outgoingMeter: Int?
    ) {
        self.companyCd = userInfo.companyCd
        self.userId = userInfo.userId
        self.allocationToken = AllocationToken(token: token)
        self.truckCd = truck.truckCd
        self.latitude = locationRecord.latitude ?? 0.0
        self.longitude = locationRecord.longitude ?? 0.0
        self.accuracy = locationRecord.accuracy ?? 0.0
        self.executeDate = locationRecord.date
        self.outgoingMeter = outgoingMeter
        self.clientInfo = clientInfo
    }
}

class AllocationToken: Codable {
    var token: String
    init(token: String) {
        self.token = token
    }
}

File: LogAndroid/TrustarApp/src/main/kotlin/jp/co/toukei/log/trustar/rest/post/UnscheduledBin.kt
package jp.co.toukei.log.trustar.rest.post

import jp.co.toukei.log.lib.util.RequestBodyJson
import jp.co.toukei.log.trustar.db.user.embedded.LocationRecord
import jp.co.toukei.log.trustar.db.user.entity.Truck
import jp.co.toukei.log.trustar.user.ClientInfo
import jp.co.toukei.log.trustar.user.UserInfo
import org.json.JSONObject
import third.jsonObj

class UnscheduledBin(
        private val truck: Truck,
        private val locationRecord: LocationRecord,
        private val clientInfo: ClientInfo,
        private val userInfo: UserInfo,
        private val outgoingMeter: Int? = null
) : RequestBodyJson {

    override fun jsonBody(): JSONObject {
        return jsonObj {
            "companyCd" v userInfo.companyCd
            "userId" v userInfo.userId

            "truckCd" v truck.truckCd

            "latitude" v (locationRecord.latitude ?: 0.0)
            "longitude" v (locationRecord.longitude ?: 0.0)
            "accuracy" v (locationRecord.accuracy ?: 0F)
            "executeDate" v locationRecord.date

            "outgoingMeter" v outgoingMeter

            "clientInfo" v clientInfo.jsonBody()
        }
    }
}
